### DISCUSSION TITLE
how reading a file?

### DISCUSSION THREAD

**Discussion by NyshimoriMizuki** [tauri-apps/tauri#7323] [Q&A] [answered]
I'm trying to read a file, almost every thing is right except for a folder called `_root_`

Why in the world the Tauri is tryin to access a `_root_` folder:  `C:\ ... \src-tauri\target\debug\_root_\data\file.txt`

tauri config:
```json
"fs": {
    "scope": [
        "$RESOURCE/**"
    ]
}
```

code in Typescript:
```typescript
const home = await readTextFile("/data/file.txt", { dir: BaseDirectory.Resource });
```

Could anyone say why it is happening??

**Comment by FabianLars** [comment]
> Why in the world the Tauri is tryin to access a _root_ folder

Because tauri reconstructs the paths of the bundled resources. so if you bundle for example `../data/file.txt` `../` will be replaced with an `_up_` alias, same for absolute paths with the root alias.
There's a helper function you can use that will get you the correct paths: <https://tauri.app/v1/api/js/path#resolveresource>

**Reply by NyshimoriMizuki** [reply]
And how uses this `resolveResource()`?

The way I tried this code and gave me the same thing
```typescript
const resPath = await resolveResource("/data/file.txt");
const home = await readTextFile(resPath);
```

**Reply by FabianLars** [reply]
In your first message it looked like you're on Windows? Did you also try it with `C:/` prefix instead of a simple `/`?

When I'm back home I'll try it on my end too to check if we have any API oversights/bugs :)

**Reply by NyshimoriMizuki** [reply]
`C:\ ... \src-tauri\target\debug\_root_\data\file.txt` is what the error message was saying that tried to access.

the `...` is a abreviation to say less and dont show unnecessary things

this is the folders structure of the project:
```
project
├ src
| └ assets
└ src-tauri
  ├ data
  ├ icons
  ├ src
  └ target
```

**Reply by FabianLars** [reply, marked as answer]
okay, so i think i misunderstood the problem then. i though the data dir is actually at the root of your filesystem. In this case you probably only need to remove the leading `/` in resolveResource to make it work.
In general, resolveResource expects path similar to how you specified them in the resources property in tauri.conf.json, a few examples:
```js
      "resources": [
        "/absolute/path/to/textfile.txt", // on windows this would start with the drive letter of course
        "relative/path/to/jsonfile.json", // relative to tauri.conf
        "../relative/in/parent/dir/file.toml" // relative to tauri.conf
      ]
// To get the paths to each of them:
await resolveResource("/absolute/path/to/textfile.txt");
await resolveResource("relative/path/to/jsonfile.json"); // -> This is what you need
await resolveResource("../relative/in/parent/dir/file.toml");
// Each will give you the full absolute path to the file, so no need for BaseDirectory
````
Again, the purpose of `resolveResource` is to turn `../` into `_up_` and `C:\\` / `/` into `_root_` nothing more.

**Reply by m4heshd** [reply]
@FabianLars Is there any way to bundle in extra resources just for the installer without this whole `_up_` shenanigans? I'd like to simply use some stuff with relative paths outside the `setup()` scope, which means without using `app.path_resolver()`. 

Is it bad practice if I manually use `_up_` in my paths? Does the behavior change on different platforms?

**Reply by FabianLars** [reply]
> Is it bad practice if I manually use _up_ in my paths? Does the behavior change on different platforms?

Imo, yes it is. No the behavior of _up_ doesn't change but the actual root dir changes iirc. Also you're less likely to have to change it in new versions if you use tauri's apis, if that matters.

>  I'd like to simply use some stuff with relative paths outside the setup() scope, which means without using app.path_resolver().

That indeed could be difficult depending on the platforms at bundle types you want to support, iirc only windows has the resources right next to the exe.

> Is there any way to bundle in extra resources just for the installer without this whole _up_ shenanigans?

In 1.5 you'll be able to use https://github.com/tauri-apps/tauri/pull/5950

**Reply by m4heshd** [reply]
> In 1.5 you'll be able to use #5950

That would be such a relief. Thank you for the response.
